Add <audio> and <video> player backends

ceyusa opened the pull request:
These patches enables audio and video playing inside Servo. 

It is bit hackish way to enable it, thus the purpose of this pull request is for an early request for comments.

It is tested with the current servo-media GStreamer backend in Linux.

~~The produced layout is not correct, since the elements after the video seems to be stacked behind, and the same with the scrolling bars.~~

~~There is no JavaScript interface yet~~, neither controls.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [X] These changes fix #6711

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/21543)
<!-- Reviewable:end -->

ceyusa committed:
dom: htmlmediaelement: add initial player bits

Also removes the usage of audio-video-metadata crate (?)

ceyusa committed:
htmlmediaelement: typo

ceyusa committed:
HACK: don't call media_source_failure_steps when eos

Because gstreamer's appsrc queue might be big enough to swallow
the whole video holding the metadata extraction until EOS.

ceyusa committed:
add a fixme message of not implemented mode

ceyusa committed:
dom: add mediaframewebrenderer

ceyusa committed:
Provide webrender_api::RenderApiSender to ScriptThread and DOM Window

This will allow the HTMLMediaElement later to get a handle to the
RenderApi for rendering video frames.

At a later time, all media handling should be moved to its own
thread/process that is communicated with via IPC. At that point this
can be removed again.

Original-patch-by: Sebastian Dröge <sebastian@centricular.com>

ceyusa committed:
layout: add media_data() method to ThreadSafeLayoutNode trait

ceyusa committed:
layout: add HTMLMediaFrameSource trait and HTMLMediaData struct

ceyusa committed:
layout: handle MediaFragmentInfo

ceyusa committed:
layout: add media_data() to ServoThreadSafeLayoutNode implementation

ceyusa committed:
layout: set building information for media fragment

ceyusa committed:
dom: add media_data() method to LayoutDom implementation

ceyusa committed:
dom: media: implement LayoutHTMLMediaElementHelpers trait for LayoutDom

ceyusa committed:
dom: framerenderer: implement HTMLMediaFrameSource trait for MediaFrameRenderer

ceyusa committed:
layout: instantiate SpecificFragmentInfo from node's media data

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @mbrubeck (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/lib.rs, components/script/Cargo.toml, components/constellation/pipeline.rs, components/script/dom/node.rs, components/script/dom/mediaframerenderer.rs and 4 more
 * @cbrewster: components/constellation/pipeline.rs
 * @kichjang: components/script/lib.rs, components/script/Cargo.toml, components/script/dom/node.rs, components/script/dom/mediaframerenderer.rs, components/script/dom/htmlmediaelement.rs and 4 more
 * @paulrouget: components/constellation/pipeline.rs
 * @emilio: components/layout/fragment.rs, components/layout/display_list/builder.rs, components/layout/construct.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify layout and script code, but no tests are modified. Please consider adding a test!

ferjm committed:
Player does not need to go inside a mutexed Arc

ferjm committed:
Do not share entire FrameRenderer with layout, only current frame

ferjm committed:
Move unsafe_no_jsmanaged_fields to trace

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21509) made this pull request unmergeable. Please resolve the merge conflicts.

ferjm committed:
Revert import style format

ferjm committed:
Clarify some FIXMEs and bail out if we fail pushing data to player

Manishearth commented:
I'm unable to get this to render at all -- can you share the testcase you're using? I'm just using [the one from W3schools](https://www.w3schools.com/html/tryit.asp?filename=tryhtml5_video) since it's very simple.

ferjm commented:
r? @nox could you take a look at the DOM changes, please?

r? @Manishearth @mrobinson could you review the layout changes, please?

Thanks!

nox commented:
r? @Manishearth

ferjm commented:
> I'm unable to get this to render at all -- can you share the testcase you're using? I'm just using the one from W3schools since it's very simple.

I already replied on IRC, but for reference, in case anyone else wants to try this, [this example](https://github.com/ferjm/web-api-tests/blob/master/video/mp4.html) should work.

ferjm commented on Cargo.toml:
```diff
@@ -27,3 +27,5 @@ opt-level = 3
 #     [patch."https://github.com/servo/<repository>"]
 #     <crate> = { path = "/path/to/local/checkout" }
+[patch."https://github.com/servo/media"]
+servo-media = { git = "https://github.com/ferjm/media", branch = "framerenderer" }
```
This should be gone once #21325 is merged

ferjm commented on components/script/script_thread.rs:
```diff
@@ -570,6 +570,9 @@ pub struct ScriptThread {
     /// The Webrender Document ID associated with this thread.
     webrender_document: DocumentId,
+
+    /// FIXME(victor):
```
@ceyusa could you elaborate more on what needs to be fixed here exactly, please?

ceyusa replied:
I used the wrong tag. It is not a fixme but a xxx, perhaps. What I mean is it would be nice to rethink this design. Perhaps it is better to create a proxy object to keep the RenderApiSender for the servo-media, rather than pass it all along from constellation to document to html element.

Manishearth replied:
I think process isolation wise script should not have access to this, cc @asajeffrey 

A proxy approach would be nice. Ideally the network task would directly talk to WR, maybe via constellation

ferjm commented on components/script_traits/lib.rs:
```diff
@@ -559,6 +561,8 @@ pub struct InitialScriptState {
     pub webvr_chan: Option<IpcSender<WebVRMsg>>,
     /// The Webrender document ID associated with this thread.
     pub webrender_document: DocumentId,
+    /// FIXME(victor): The Webrender API sender in this constellation's pipeline
```
@ceyusa what needs to be fixed here?

ceyusa replied:
The same I commented above, not a fixme but a request to rethink this approach,

ferjm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -602,6 +690,7 @@ impl HTMLMediaElement {
                     .. RequestInit::default()
                 };
 
+                self.setup_media_player();
```
I suspect that this is not the best place to setup the media player. I need to properly read the spec, but I would say that we should be setting it up a bit earlier, and ideally handling any potential setup failure.

ferjm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -269,6 +352,9 @@ impl HTMLMediaElement {
                         this.upcast::<EventTarget>().fire_event(atom!("pause"));
 
+                        //FIXME(victor)
+                        //this.player.pause();
```
This can be uncommented once we update servo-media to its latest revision (after #21325 lands)

Manishearth commented:
Hm, so it's not that they're stacked behind, it's that they seem to get laid out as 0x0 boxes at (0,0). Curious.

ferjm committed:
Abort in-progress fetching process during load algorithm

ferjm committed:
Add show poster flag

ferjm committed:
Player pause

Manishearth reviewed (approved):
The DOM side approach looks okay, though we might want to rethink the sender stuff.

Any progress on figuring out what's going on with layout? cc @mbrubeck @mrobinson

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21723) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21737) made this pull request unmergeable. Please resolve the merge conflicts.

ferjm committed:
Cancel fetch in some media data processing steps

ferjm committed:
videoWidth and videoHeight params

ferjm committed:
Media element duration param

ferjm committed:
Rustfmt and make tidy happy

ferjm committed:
Set input size

jdm commented:
@Manishearth Can you review the layout changes as well?

ferjm commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 40a1fe74cd8d86b260ff570179469b73d287d189 with merge 6178383f62e76e2536d765d3301cba13fd9978c0...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9913)

ferjm committed:
Do not set metadata until we have enough data

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9930)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21794) made this pull request unmergeable. Please resolve the merge conflicts.

Manishearth reviewed (approved):
r=me on both the layout and DOM code

Manishearth commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -18,36 +17,113 @@ use dom::bindings::error::{Error, ErrorResult};
+    api: RenderApi,
+    current_frame: Option<(ImageKey, i32, i32)>,
+    old_frame: Option<ImageKey>,
+    very_old_frame: Option<ImageKey>,
```
some doc comments on why this is necessary?

ferjm committed:
Fix HAVE_METADATA and HAVE_CURRENT_DATA state transitions

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 37ff37bd2d3db535172b7643cb755a12a0d28d1a with merge f3f819a3dab8fe5bdb8bf2611f7bf27dfb114bd8...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9940)

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit edf7ccf6277c508c12843d360ce60a2846e61db9 with merge 249569211953e28d3cea1a5503a556effc25537a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9943)

ferjm commented:
Most of these tests are passing locally, so I suspect there's something wrong with the GStreamer setup on CI. Let's see if GST_DEBUG shows something useful.

@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 0eba93abafa3c4726c08e96dca927de32a65dbce with merge 0b89a62c43270ec994131e995dd9071d621587cf...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9944)

ferjm commented:
There you go: `Missing decoder: Ogg (video/ogg)`

ceyusa committed:
Network EOF does't imply media processing EOS

Since media processing is highly asynchronous, there is no need to
match one event with the other. Then error handling is done when
the player emits the EOS event.

The have_metadata attribute is not required anymore.

ferjm commented:
We have new GStreamer binaries for CI that should include the missing decoders, so let's try again.

@bors-servo retry

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 1c73c720af71aa304eefb286a2816f72bf334277 with merge ccff73c6340c9201d534d90c8fc571336e02149e...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9976)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 1c73c720af71aa304eefb286a2816f72bf334277 with merge 6e3f5a980205d0bff997dba97821c49e41f19c1c...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9977)

jdm commented:
Ha, it switched from complaining about ogg decoders to complaining about theora decoders.

ferjm commented:
Ugh. Well, that's some progress at least. I'll add theora decoders to the binaries as well.

ferjm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 1c73c720af71aa304eefb286a2816f72bf334277 with merge ebc9c1e715c5b62e6f9993fa944b7de1a9c52473...

CYBAI commented:
```
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/2dcontext/imagebitmap/createImageBitmap-invalid-args.html", "line": 38110, "action": "test_result", "expected": "TIMEOUT"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/2dcontext/imagebitmap/createImageBitmap-origin.sub.html", "line": 38112, "action": "test_result", "expected": "OK"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/2dcontext/imagebitmap/canvas-createImageBitmap-video-resize.html", "line": 43263, "action": "test_result", "expected": "OK"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/2dcontext/imagebitmap/createImageBitmap-transfer.html", "line": 55185, "action": "test_result", "expected": "OK"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "setting src attribute on non-autoplay audio should trigger loadedmetadata event", "test": "/html/semantics/embedded-content/media-elements/event_loadedmetadata_noautoplay.html", "line": 101876, "action": "test_result", "expected": "PASS"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "setting src attribute on non-autoplay video should trigger loadedmetadata event", "test": "/html/semantics/embedded-content/media-elements/event_loadedmetadata_noautoplay.html", "line": 101878, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/event_loadedmetadata_noautoplay.html", "line": 101879, "action": "test_result", "expected": "OK"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/loading-the-media-resource/load-events-networkState.html", "line": 103109, "action": "test_result", "expected": "OK"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "audio.readyState should be >= HAVE_CURRENT_DATA during loadeddata event", "test": "/html/semantics/embedded-content/media-elements/readyState_during_loadeddata.html", "line": 110306, "action": "test_result", "expected": "PASS"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "video.readyState should be >= HAVE_CURRENT_DATA during loadeddata event", "test": "/html/semantics/embedded-content/media-elements/readyState_during_loadeddata.html", "line": 110308, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/readyState_during_loadeddata.html", "line": 110309, "action": "test_result", "expected": "OK"}
{"status": "FAIL", "group": "default", "message": "assert_equals: expected 0 but got 4", "stack": "v.onerror<@http://web-platform.test:8000/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-insert-into-iframe.html:12:5\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1561:20\nTest.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1585:20\n", "subtest": "NOT invoking resource selection by inserting into other document with src set", "test": "/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-insert-into-iframe.html", "line": 111453, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-insert-into-iframe.html", "line": 111454, "action": "test_result", "expected": "OK"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/playing-the-media-resource/pause-move-within-document.html", "line": 112020, "action": "test_result", "expected": "OK"}
{"status": "FAIL", "group": "default", "message": "assert_equals: networkState in onerror expected 3 but got 1", "stack": "t</v.onerror<@http://web-platform.test:8000/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-remove-from-document-networkState.html:18:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1561:20\nTest.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1585:20\n", "subtest": "NOT invoking resource selection with implicit pause() when networkState is not NETWORK_EMPTY", "test": "/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-remove-from-document-networkState.html", "line": 120867, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-remove-from-document-networkState.html", "line": 120868, "action": "test_result", "expected": "OK"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "setting src attribute on autoplay audio should trigger loadedmetadata event", "test": "/html/semantics/embedded-content/media-elements/event_loadedmetadata.html", "line": 125000, "action": "test_result", "expected": "PASS"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "setting src attribute on autoplay video should trigger loadedmetadata event", "test": "/html/semantics/embedded-content/media-elements/event_loadedmetadata.html", "line": 125002, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/event_loadedmetadata.html", "line": 125003, "action": "test_result", "expected": "OK"}
{"status": "FAIL", "group": "default", "message": "assert_equals: networkState in onerror expected 3 but got 1", "stack": "t</v.onerror<@http://web-platform.test:8000/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-pause-networkState.html:17:7\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1561:20\nTest.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1585:20\n", "subtest": "NOT invoking resource selection with pause() when networkState is not NETWORK_EMPTY", "test": "/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-pause-networkState.html", "line": 130219, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/loading-the-media-resource/resource-selection-invoke-pause-networkState.html", "line": 130220, "action": "test_result", "expected": "OK"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/playing-the-media-resource/pause-move-to-other-document.html", "line": 131756, "action": "test_result", "expected": "OK"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "audio.readyState should be >= HAVE_METADATA during loadedmetadata event", "test": "/html/semantics/embedded-content/media-elements/readyState_during_loadedmetadata.html", "line": 131760, "action": "test_result", "expected": "PASS"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "video.readyState should be >= HAVE_METADATA during loadedmetadata event", "test": "/html/semantics/embedded-content/media-elements/readyState_during_loadedmetadata.html", "line": 131762, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/readyState_during_loadedmetadata.html", "line": 131763, "action": "test_result", "expected": "OK"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/mozilla/focus_blur.html", "line": 147999, "action": "test_result", "expected": "OK"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "setting src attribute on autoplay audio should trigger loadeddata event", "test": "/html/semantics/embedded-content/media-elements/event_loadeddata.html", "line": 186420, "action": "test_result", "expected": "PASS"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "setting src attribute on autoplay video should trigger loadeddata event", "test": "/html/semantics/embedded-content/media-elements/event_loadeddata.html", "line": 186422, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/event_loadeddata.html", "line": 186423, "action": "test_result", "expected": "OK"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "setting src attribute on autoplay audio should trigger loadedmetadata then loadeddata event", "test": "/html/semantics/embedded-content/media-elements/event_order_loadedmetadata_loadeddata.html", "line": 193493, "action": "test_result", "expected": "PASS"}
{"status": "NOTRUN", "group": "default", "message": null, "stack": null, "subtest": "setting src attribute on autoplay video should trigger loadedmetadata then loadeddata event", "test": "/html/semantics/embedded-content/media-elements/event_order_loadedmetadata_loadeddata.html", "line": 193495, "action": "test_result", "expected": "PASS"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/event_order_loadedmetadata_loadeddata.html", "line": 193496, "action": "test_result", "expected": "OK"}
{"status": "TIMEOUT", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/loading-the-media-resource/autoplay-overrides-preload.html", "line": 200791, "action": "test_result", "expected": "OK"}
```

ferjm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 1c73c720af71aa304eefb286a2816f72bf334277 with merge 6c924b50f01a4f3debf03aed2e2ade363d81e9de...

ferjm commented:
The failing innerText related tests are caused because we don't support child sources yet #21481.

ferjm committed:
Fix Travis complains about lock file

ferjm committed:
Move player setup to common path, deal with setup error and clean debug msgs

ferjm committed:
Update servo-media: handle decoder errors and use safer/nicer player API

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit e6a3bf067bda6d250d67deaed671bd235bbd56ba with merge cc9e0fefda370cd931071a8a01cf39f2c5f3bba9...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9999)

ferjm committed:
Handle Player errors

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 6c821b6d88aa081807987efdeff4e7e340ccb326 with merge 7b1010c5339855ba973aa3e00c65b21d4dad24e8...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10008)

ferjm commented:
There's still something fishy with the player setup on the servo-media side that causes these WPTs errors :\

ferjm committed:
Make sure we do not skip the HaveMetadata state. Fixes a bunch of tests

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 572d9a850ee3cb6d3239fd94ad1b20bed5cc3ee8 with merge 145393e51af4972751a01a106f45afb83e1000f1...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/10035)

ferjm committed:
Update tests expectations

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 5071011131bd8f66b332040821f707d196b040dc with merge e801b018465cc9331085409f26ee0580864a8a7d...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/10040)

ferjm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 5071011131bd8f66b332040821f707d196b040dc with merge 8247333581cf3ab3d69f6c3b20969648f653bea9...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/10042)

ferjm commented:
@bors-servo retry
* HTTPError: HTTP Error 500: INTERNAL SERVER ERROR

bors-servo commented:
:hourglass: Trying commit 5071011131bd8f66b332040821f707d196b040dc with merge 41cc952dc4abcbd51dc006a5934a378ad546360b...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10024)

ferjm commented:
> HTTPError: HTTP Error 500: INTERNAL SERVER ERROR

@jdm is the intermittent tracker tool broken?

jdm commented:
Yes. Follow https://github.com/servo/intermittent-failure-tracker/issues/5 to reset it. I am getting on a plane :(

ferjm commented:
Ok, thanks!

> I am getting on a plane :(

Don't be sad, planes are cool :P. Have a safe flight :)

ferjm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 5071011131bd8f66b332040821f707d196b040dc with merge f42240e819a030f2e8e7be1a151625352b853871...

jdm commented:
Even though it's going to report that buildbot timed out for some reason, the all the tests passed.

jdm commented:
@bors-servo r-

bors-servo commented:
:boom: Test timed out

ferjm commented:
@bors-servo r=Manishearth

bors-servo commented:
:pushpin: Commit 5071011 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 5071011131bd8f66b332040821f707d196b040dc -->

bors-servo commented:
:hourglass: Testing commit 5071011131bd8f66b332040821f707d196b040dc with merge 02c982b1ff1957e67982b8f3dd143ae19aa77fc2...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/12713)

ferjm commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android-x86](http://build.servo.org/builders/android-x86/builds/828), [arm32](http://build.servo.org/builders/arm32/builds/12897), [arm64](http://build.servo.org/builders/arm64/builds/12936), [linux-dev](http://build.servo.org/builders/linux-dev/builds/13008), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10032), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/10052), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/13031), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5492), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9751), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9801), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5557), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/RM-p881iTsaxPl3mDw8WaQ), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8392) are reusable. Rebuilding only [android](http://build.servo.org/builders/android/builds/12713), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5486), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5492)...

CYBAI commented:
```
{"status": "FAIL", "group": "default", "message": "assert_unreached: Reached unreachable code", "stack": null, "subtest": "calling play() on a sufficiently long audio should trigger timeupdate event", "test": "/html/semantics/embedded-content/media-elements/event_timeupdate_noautoplay.html", "line": 67449, "action": "test_result", "expected": "NOTRUN"}
{"status": "FAIL", "group": "default", "message": "assert_unreached: Reached unreachable code", "stack": null, "subtest": "calling play() on a sufficiently long video should trigger timeupdate event", "test": "/html/semantics/embedded-content/media-elements/event_timeupdate_noautoplay.html", "line": 67451, "action": "test_result", "expected": "NOTRUN"}
{"status": "OK", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/event_timeupdate_noautoplay.html", "line": 67452, "action": "test_result", "expected": "TIMEOUT"}
```

jdm commented:
That test passed on linux, so it may be a new intermittent result.

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5494)

ferjm commented:
The test should be a TIMEOUT for now because we are not sending the expected `timeupdate` event yet. It seems that on linux we get the TIMEOUT, but on mac we get an error event before timing out. So this might be a platform dependent issue (i.e. missing codecs on the mac builder?). I see that we are [installing `gst-plugins-base` on mac](https://github.com/servo/saltfs/blob/master/servo-build-dependencies/init.sls#L33), but I don't see us specifying the required options [to build with codecs support](https://formulae.brew.sh/formula/gst-plugins-base) (--with-libogg, --with-libvorbis, --with-theora, etc.).

CC @Manishearth

ferjm commented:
I've confirmed that there's no ogg, vorbis or theora support on the mac builders. For example:

```
administrators-Mac-mini:~ root# gst-inspect-1.0 | grep ogg
typefindfunctions: application/x-id3v2: mp3, mp2, mp1, mpga, ogg, flac, tta
typefindfunctions: application/x-id3v1: mp3, mp2, mp1, mpga, ogg, flac, tta
typefindfunctions: application/ogg: ogg, oga, ogv, ogm, ogx, spx, anx, axa, axv
typefindfunctions: application/x-ogg-skeleton: no extensions
```

Should be something like:

```
❯ gst-inspect-1.0 | grep ogg
ogg:  oggdemux: Ogg demuxer
ogg:  oggmux: Ogg muxer
ogg:  ogmaudioparse: OGM audio stream parser
ogg:  ogmvideoparse: OGM video stream parser
ogg:  ogmtextparse: OGM text stream parser
ogg:  oggparse: Ogg parser
ogg:  oggaviparse: Ogg AVI parser
typefindfunctions: application/x-ogg-skeleton: no extensions
typefindfunctions: application/ogg: ogg, oga, ogv, ogm, ogx, spx, anx, axa, axv
typefindfunctions: application/x-id3v1: mp3, mp2, mp1, mpga, ogg, flac, tta
typefindfunctions: application/x-id3v2: mp3, mp2, mp1, mpga, ogg, flac, tta
libav:  avmux_ogg: libav Ogg muxer (not recommended, use oggmux instead)
```

jdm commented:
I'm investigating how to install the packages on mac with the right options. https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.mac_brew.html exists but I don't know how to use it from the SLS files.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21878) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
https://github.com/servo/saltfs/pull/898 is my attempt to fix the situation but it is hitting problems.

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/13057)

Manishearth commented:
fwiw I'm re-highstating right now, so this may break.

jdm commented:
Oops.

Manishearth commented:
@bors-servo try retry

bors-servo commented:
:hourglass: Trying commit 92a10e2616c53a688b389098928af81f03173ea0 with merge 9bd859db28d072d4dba6d137869e78047b0e1343...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9779)

Manishearth commented:
Something's funky with mac5

Manishearth commented:
@bors-servo retry


took it offline, it's got its own problems

bors-servo commented:
:hourglass: Trying commit 92a10e2616c53a688b389098928af81f03173ea0 with merge 34eaebf6e26d91ff53610d5b6d5af9ab20d00e67...

Manishearth commented:
somehow the virtualenv needed a reinstall :man_shrugging:

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5524)

Manishearth commented:
```
{"status": "FAIL", "group": "default", "message": "assert_unreached: Reached unreachable code", "stack": null, "subtest": "calling play() on a sufficiently long audio should trigger timeupdate event", "test": "/html/semantics/embedded-content/media-elements/event_timeupdate_noautoplay.html", "line": 65788, "action": "test_result", "expected": "NOTRUN"}
{"status": "FAIL", "group": "default", "message": "assert_unreached: Reached unreachable code", "stack": null, "subtest": "calling play() on a sufficiently long video should trigger timeupdate event", "test": "/html/semantics/embedded-content/media-elements/event_timeupdate_noautoplay.html", "line": 65790, "action": "test_result", "expected": "NOTRUN"}
{"status": "OK", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/media-elements/event_timeupdate_noautoplay.html", "line": 65791, "action": "test_result", "expected": "TIMEOUT"}
```

legit failure, we don't support that event.

jdm commented:
Those failures don't happen on linux. Fernando previously tracked those failures down to the lack of the codecs on mac.

Manishearth commented:
I _suspect_ this is brew not being smart about reinstalls when you're installing with extra args. May need to do an uninstall before resalting (but first I'll verify that this is indeed what's happening)

ferjm committed:
Disable event_timeupdate_noautoplay test on mac

ferjm commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 7e6661d9abd7b54b14853fbb469b5d7fa4442b84 with merge 3b153af49c8635e01b3a8eed86b5a1f901d58353...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5589)

ferjm commented:
```
  ▶ CRASH [expected OK] /html/dom/elements/the-innertext-idl-attribute/getter.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.1.0-devel
  │ called `Result::unwrap()` on an `Err` value: ChannelClosedError (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(2), index: PipelineIndex(7) }, at libcore/result.rs:1009)
  │ 2018-10-09 05:04:32.098 servo[85591:104129169] Metadata.framework [Error]: couldn't get the client port
  │ stack backtrace:
  │    0:        0x1069d117e - __ZN9backtrace9backtrace5trace17h7e39d6bd7c678f2eE
  │    1:        0x1069d1e8c - __ZN72_$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$7default17ha5434a657e6f38c1E
  │    2:        0x1069d1f0d - __ZN9backtrace7capture9Backtrace3new17h027ac9deb1d86dcfE
  │    3:        0x103ebc5f2 - __ZN5servo4main28_$u7b$$u7b$closure$u7d$$u7d$17he67dbfce30c0cf9fE
  │    4:        0x1069f6838 - __ZN3std9panicking20rust_panic_with_hook17h14490a1041ee4235E
  │    5:        0x1069f634c - __ZN3std9panicking18continue_panic_fmt17h7beddad0ba45f4eaE
  │    6:        0x1069f6238 - _rust_begin_unwind
  │    7:        0x106a13d01 - __ZN4core9panicking9panic_fmt17hc2d37e7312d32a42E
  │    8:        0x104a5de6b - __ZN4core6result13unwrap_failed17hde0251e98d282c5cE
  │    9:        0x1047f520b - __ZN6script3dom8document8Document15set_quirks_mode17hb0e4b28a660531a2E
  │   10:        0x104aa72bd - __ZN71_$LT$html5ever..tree_builder..TreeBuilder$LT$Handle$C$$u20$Sink$GT$$GT$4step17h5b5c27613eac3c68E
  │   11:        0x104a17fb3 - __ZN125_$LT$html5ever..tree_builder..TreeBuilder$LT$Handle$C$$u20$Sink$GT$$u20$as$u20$html5ever..tokenizer..interface..TokenSink$GT$13process_token17heb4ee21f941c8e22E
  │   12:        0x104d536b2 - __ZN52_$LT$html5ever..tokenizer..Tokenizer$LT$Sink$GT$$GT$13process_token17he4a887dc1ccb4b8eE.llvm.7703268766650612569
  │   13:        0x104d5f142 - __ZN52_$LT$html5ever..tokenizer..Tokenizer$LT$Sink$GT$$GT$4step17h3f8979140582a039E
  │   14:        0x104d5afca - __ZN52_$LT$html5ever..tokenizer..Tokenizer$LT$Sink$GT$$GT$3run17hd00c861be46bd011E.llvm.7703268766650612569
  │   15:        0x104e82eeb - __ZN6script3dom11servoparser4html9Tokenizer4feed17h112165b7aed7ca0eE
  │   16:        0x10479d31c - __ZN6script3dom11servoparser11ServoParser13do_parse_sync17h0abe13de0b4e0d20E
  │   17:        0x104a307dd - __ZN14profile_traits4time7profile17h56bd06f49e294cf3E
  │   18:        0x10479d0d6 - __ZN6script3dom11servoparser11ServoParser10parse_sync17h40ec5210842cd6d2E.llvm.15457495438227833469
  │   19:        0x10479fee8 - __ZN93_$LT$script..dom..servoparser..ParserContext$u20$as$u20$net_traits..FetchResponseListener$GT$22process_response_chunk17h97bc4cee11ab4773E
  │   20:        0x104904ce5 - __ZN6script13script_thread12ScriptThread29handle_msg_from_constellation17hf41fce0ec6ee0376E
  │   21:        0x1048fe194 - __ZN6script13script_thread12ScriptThread11handle_msgs17hf1278d1c475732c2E.llvm.5337641549442363273
  │   22:        0x104a3ccf7 - __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h5cbb6512ca403ff6E
  │   23:        0x10429cfad - __ZN3std9panicking3try7do_call17h63359dbf63ea5718E.llvm.7079729101609438744
  │   24:        0x106a02dae - ___rust_maybe_catch_panic
  │   25:        0x104d52bf6 - __ZN50_$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$8call_box17hcc1a52b90f7d7a98E
  │   26:        0x1069da3eb - __ZN3std3sys4unix6thread6Thread3new12thread_start17hba53abd905cebc29E
  │   27:     0x7fffa187799c - __pthread_body
  │   28:     0x7fffa1877919 - __pthread_start
  │ ERROR 2018-10-09T09:04:32Z: servo: called `Result::unwrap()` on an `Err` value: ChannelClosedError
  └ Pipeline failed in hard-fail mode.  Crashing!
```

The crash looks unrelated to this PR.

ferjm commented:
@bors-servo r=Manishearth

bors-servo commented:
:pushpin: Commit 7e6661d has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 7e6661d9abd7b54b14853fbb469b5d7fa4442b84 -->

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/12744), [android-x86](http://build.servo.org/builders/android-x86/builds/858), [arm32](http://build.servo.org/builders/arm32/builds/12926), [arm64](http://build.servo.org/builders/arm64/builds/12965), [linux-dev](http://build.servo.org/builders/linux-dev/builds/13038), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10069), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/10091), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/13061), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5520), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5524), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9782), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9833), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5526), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/c_pN8MYrTGelPoHdngDT1Q), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8421) are reusable. Rebuilding only [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5589)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/12744), [android-x86](http://build.servo.org/builders/android-x86/builds/858), [arm32](http://build.servo.org/builders/arm32/builds/12926), [arm64](http://build.servo.org/builders/arm64/builds/12965), [linux-dev](http://build.servo.org/builders/linux-dev/builds/13038), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10069), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/10091), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/13061), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5520), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5524), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9782), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9833), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5590), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5526), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/c_pN8MYrTGelPoHdngDT1Q), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8421)
Approved by: Manishearth
Pushing 3b153af49c8635e01b3a8eed86b5a1f901d58353 to master...
